base/abstractarray.jl

```diff
@@ -1048,6 +1048,50 @@ function _setindex!(::IndexCartesian, A::AbstractArray, v, I::Vararg{Int,M}) whe
     r
 end
 
+## rudimentary aliasing detection ##
+"""
+    unalias(dest, A)
+
+Return either A or a copy of A, in a rough effort to prevent modifications to `dest` from
+affecting the returned object. No guarantees are provided, and each custom array must
+opt-into aliasing detection by overloading this method by specializing on the second argument.
+
+This function must return an object of exactly the same type as A for performance and type-stability.
+
+See also `mightalias` and `dataids`.
+"""
+unalias(dest, A) = A
+unalias(dest, A::Array) = mightalias(dest, A) ? copy(A) : A
+@inline unalias(dest, As::Tuple) = (unalias(dest, As[1]), unalias(dest, tail(As))...)
```

Comment:
Why not use varargs?

Reply:
Why not?  I liked this symmetry in the calling conventions:

```
function setindex(dest, source, I...)
    source′ = unalias(dest, source)
    I′ = unalias(dest, I)
```

It's also nice in that the tuple case mostly fits the explanation above without needing to call it out — it'll either return exactly the tuple or a (partial) copy.

Reply:
Yeah, I guess you need to choose between symmetry and consistency with `setindex`.